Remove FunctionSpace and Discretization

kohr-h opened the pull request:
**Rough Overview:**
- `Discretization` and `DiscretizationElement` are gone, functionality has been moved onto `DiscreteLp` and `DiscreteLpElement`
- `FunctionSpace` and `FunctionSpaceElement` are gone, functionality for vectorization and evaluation has been broken into a bunch of functions that live in `odl.discr.discr_utils`.
  For instance, `FunctionSpace.element()` created a wrapper that unified the interface of functions with and without `out` argument. This is done in the function `make_func_for_sampling`.
- Support for non-vectorized functions in `DiscreteLp.element()` is removed.
  *Rationale:* Why did we ever support this?
- Some parts of `discretization.py` (now gone) are inlined in `lp_discr.py`, which makes it appear as if it had changed a lot. It hasn't.
- `discr_sequence_space` is gone.
  *Rationale:* lack of usefulness.

- ~`FunctionSpaceMapping` classes have simply been turned into functions and moved over to the `discr_utils` module~ done in #1520 
- ~`interp` is no longer a property of the space but up to each operator (which I consider to be the better approach)~ done in #1520 

Closes #1457

pep8speaks commented:
Checking new PR...







 - In the file [`odl/discr/discr_ops.py`](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/discr_ops.py), following are the PEP8 issues :

> [Line 480:71](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/discr_ops.py#L480): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 321:49](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/discr_ops.py#L321): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 320:46](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/discr_ops.py#L320): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator

 - Complete extra results for this file :

>                 if (ran.is_uniform_byaxis[i] and                                             ^                    domain.is_uniform_byaxis[i] and                                                ^        nodes_on_bdry = ([(bool(nodes_on_bdry), bool(nodes_on_bdry))] *                                                                      ^---

 - In the file [`odl/discr/discr_utils.py`](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/discr_utils.py), following are the PEP8 issues :

> [Line 555:53](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/discr_utils.py#L555): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 405:37](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/discr_utils.py#L405): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator

 - Complete extra results for this file :

>         if (schemes[i] == 'nearest' and                                    ^            norm_distances.append((xi - cvec[idcs]) /                                                    ^---

 - In the file [`odl/discr/lp_discr.py`](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/lp_discr.py), following are the PEP8 issues :

> [Line 1460:72](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/lp_discr.py#L1460): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 1459:45](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/lp_discr.py#L1459): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 1440:43](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/lp_discr.py#L1440): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 1278:65](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/lp_discr.py#L1278): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 1233:17](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/lp_discr.py#L1233): [E117](https://duckduckgo.com/?q=pep8%20E117) over-indented
> [Line 1231:33](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/lp_discr.py#L1231): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 716:37](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/lp_discr.py#L716): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 621:1](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/lp_discr.py#L621): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (3)
> [Line 589:68](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/lp_discr.py#L589): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 583:68](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/lp_discr.py#L583): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 578:32](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/lp_discr.py#L578): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 577:47](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/lp_discr.py#L577): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 478:55](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/lp_discr.py#L478): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 255:47](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/lp_discr.py#L255): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 254:46](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/discr/lp_discr.py#L254): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator

 - Complete extra results for this file :

>                 np.allclose(bdry_fracs, 1.0) or                                             ^                self.exponent == float('inf') or                                              ^                super(DiscreteLp, self).__eq__(other) and                                                      ^                self.exponent == float('inf') or                                              ^                self.ndim == 0 or                               ^                    not isinstance(self.weighting, ConstWeighting) or                                                                   ^                    not isinstance(self.weighting, ConstWeighting) or                                                                   ^class DiscreteLpElement(Tensor):^        return (other in self.space and                                    ^        elif (method == 'outer' and                                ^                raise TypeError(                ^                        weighting = (inp1.space.weighting.const *                                                                ^            indices = ((slice(None),) * 2 +                                          ^            indices = (tuple(indices[:pos]) +                                            ^                       (slice(None),) * (self.ndim - len(indices) + 1) +                                                                       ^---

 - In the file [`odl/operator/pspace_ops.py`](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/operator/pspace_ops.py), following are the PEP8 issues :

> [Line 1138:52](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/operator/pspace_ops.py#L1138): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 1137:33](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/operator/pspace_ops.py#L1137): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 947:52](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/operator/pspace_ops.py#L947): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 946:33](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/operator/pspace_ops.py#L946): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 772:52](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/operator/pspace_ops.py#L772): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 771:33](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/operator/pspace_ops.py#L771): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 480:65](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/operator/pspace_ops.py#L480): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator

 - Complete extra results for this file :

>             linear_index = np.flatnonzero((self.ops.row == row) &                                                                ^        if (len(operators) == 2 and                                ^                isinstance(operators[0], Operator) and                                                   ^        if (len(operators) == 2 and                                ^                isinstance(operators[0], Operator) and                                                   ^        if (len(operators) == 2 and                                ^                isinstance(operators[0], Operator) and                                                   ^---

 - In the file [`odl/operator/tensor_ops.py`](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/operator/tensor_ops.py), following are the PEP8 issues :

> [Line 929:28](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/operator/tensor_ops.py#L929): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 928:75](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/operator/tensor_ops.py#L928): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 835:45](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/operator/tensor_ops.py#L835): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 621:62](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/operator/tensor_ops.py#L621): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 421:59](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/operator/tensor_ops.py#L421): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 407:47](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/operator/tensor_ops.py#L407): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 205:46](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/operator/tensor_ops.py#L205): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 192:63](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/operator/tensor_ops.py#L192): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator

 - Complete extra results for this file :

>                 self.__weights = (self.domain.weighting.const *                                                              ^            if (not np.all(self.weights > 0) or                                             ^        if (vfspace.field == ComplexNumbers() and                                              ^                self.__weights = (vfspace.weighting.const *                                                          ^            self.__ran_weights = (self.range.weighting.const *                                                             ^            if (range_shape != domain.shape and                                            ^            elif (parse_version(np.__version__) < parse_version('1.13.0') and                                                                          ^                  x is out and                           ^---

 - In the file [`odl/phantom/geometric.py`](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/phantom/geometric.py), following are the PEP8 issues :

> [Line 88:30](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/phantom/geometric.py#L88): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator

 - Complete extra results for this file :

>             result = (result &                             ^---

 - In the file [`odl/phantom/noise.py`](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/phantom/noise.py), following are the PEP8 issues :

> [Line 142:19](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/phantom/noise.py#L142): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\m'

 - Complete extra results for this file :

>     :math:`k \\in \mathbb{N}_0` is given by                  ^---



 - In the file [`odl/tomo/operators/ray_trafo.py`](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/tomo/operators/ray_trafo.py), following are the PEP8 issues :

> [Line 217:68](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/tomo/operators/ray_trafo.py#L217): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 165:62](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/tomo/operators/ray_trafo.py#L165): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator

 - Complete extra results for this file :

>             if (isinstance(geometry, Parallel3dAxisGeometry) and                                                             ^            elif (isinstance(reco_space.weighting, ConstWeighting) and                                                                   ^---

 - In the file [`odl/trafos/fourier.py`](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/trafos/fourier.py), following are the PEP8 issues :

> [Line 619:56](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/trafos/fourier.py#L619): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 463:72](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/trafos/fourier.py#L463): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator

 - Complete extra results for this file :

>                 return (np.prod(np.take(self.domain.shape, self.axes)) *                                                                       ^                return (np.fft.fftn(x, axes=self.axes) /                                                       ^---



 - In the file [`odl/util/testutils.py`](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/util/testutils.py), following are the PEP8 issues :

> [Line 339:17](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/util/testutils.py#L339): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> [Line 263:17](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/util/testutils.py#L263): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator

 - Complete extra results for this file :

>                 and p.name == 'skipif'                ^                + 1j * np.random.randn(*space.shape)                ^---

 - In the file [`odl/util/utility.py`](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/util/utility.py), following are the PEP8 issues :

> [Line 1324:60](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/util/utility.py#L1324): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 1218:56](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/util/utility.py#L1218): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 1163:60](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/util/utility.py#L1163): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 1162:34](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/util/utility.py#L1162): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 1112:59](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/util/utility.py#L1112): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 1110:59](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/util/utility.py#L1110): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 939:61](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/util/utility.py#L939): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 938:34](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/util/utility.py#L938): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 908:57](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/util/utility.py#L908): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 907:32](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/odl/util/utility.py#L907): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator

 - Complete extra results for this file :

>         elif (np.isscalar(arg) and                               ^              np.array(arg).real.astype('int64') != arg and                                                        ^        elif (np.isscalar(value) and                                 ^              np.array(value).real.astype('int64') != value and                                                            ^    pos_sig_len = (sum(len(pstr) for pstr in pos_strings) +                                                          ^    opt_sig_len = (sum(len(pstr) for pstr in opt_strings) +                                                          ^        if (not allow_mixed_seps or                                 ^                any('\n' in s for s in [pos_str, opt_str]) or                                                           ^    if (len(inst_str) + 1 + len(attr_str) <= linewidth or                                                       ^    if (len(inst_str) + 1 + len(meth_str) + 1 <= linewidth or                                                           ^---

 - In the file [`setup.py`](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/setup.py), following are the PEP8 issues :

> [Line 101:80](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/setup.py#L101): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (232 > 79 characters)
> [Line 100:80](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/setup.py#L100): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (170 > 79 characters)
> [Line 99:80](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/setup.py#L99): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (163 > 79 characters)
> [Line 98:80](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/setup.py#L98): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (121 > 79 characters)
> [Line 97:80](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/setup.py#L97): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (139 > 79 characters)
> [Line 96:80](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/setup.py#L96): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (176 > 79 characters)
> [Line 95:80](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/setup.py#L95): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 87:80](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/setup.py#L87): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (116 > 79 characters)
> [Line 86:80](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/setup.py#L86): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 79 characters)
> [Line 85:80](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/setup.py#L85): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (120 > 79 characters)
> [Line 82:80](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/setup.py#L82): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (199 > 79 characters)
> [Line 81:80](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/setup.py#L81): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (199 > 79 characters)
> [Line 79:80](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/setup.py#L79): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (331 > 79 characters)
> [Line 63:22](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/setup.py#L63): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> [Line 62:17](https://github.com/odlgroup/odl/blob/cf14f1fec165df0a4c2ac9ac1c223f31a5cbfd94/setup.py#L62): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator

 - Complete extra results for this file :

>                 and (basename.startswith('test_')                ^                     or basename.endswith('_test'))                     ^Operator Discretization Library (ODL) is a Python library for fast prototyping focusing on (but not restricted to) inverse problems. ODL is being developed at `KTH Royal Institute of Technology, Stockholm <https://www.kth.se/en/sci/institutioner/math>`_, and `Centrum Wiskunde & Informatica (CWI), Amsterdam <https://www.cwi.nl>`_.                                                                               ^The main intent of ODL is to enable mathematicians and applied scientists to use different numerical methods on real-world problems without having to implement all necessary parts from the bottom up.                                                                               ^This is reached by an `Operator` structure which encapsulates all application-specific parts, and a high-level formulation of solvers which usually expect an operator, data and additional parameters.                                                                               ^1. Different problems can be solved with the same method (e.g. TV regularization) by simply switching operator and data.                                                                               ^2. The same problem can be solved with different methods by simply calling into different solvers.                                                                               ^3. Solvers and application-specific code need to be written only once, in one place, and can be tested individually.                                                                               ^- Efficient and well-tested data containers based on Numpy (default) or CUDA (optional)                                                                               ^- Objects to represent mathematical notions like vector spaces and operators, including properties as expected from mathematics (inner product, norm, operator composition, ...)                                                                               ^- Convenience functionality for operators like arithmetic, composition, operator matrices etc., which satisfy the known mathematical rules.                                                                               ^- Out-of-the-box support for frequently used operators like scaling, partial derivative, gradient, Fourier transform etc.                                                                               ^- A versatile and pluggable library of optimization routines for smooth and non-smooth problems, such as CGLS, BFGS, Chambolle-Pock and Douglas-Rachford splitting.                                                                               ^- Support for tomographic imaging with a unified geometry representation and bindings to external libraries for efficient computation of projections and back-projections.                                                                               ^- Standardized tests to validate implementations against expected behavior of the corresponding mathematical object, e.g. if a user-defined norm satisfies `norm(x + y) <= norm(x) + norm(y)` for a number of input vectors `x` and `y`.                                                                               ^---

kohr-h commented:
Maybe it would already now be good to have some input. From the point of view of functionality, this PR is complete. A big part of it is just moving code around and wiring it up again, so I don't think it's that much to review after all.

kohr-h commented:
All tests are fixed now. I haven't had the opportunity to check ASTRA CUDA yet, but there were no changes, so it should work. Will look at it soon.

Anyway, now the PR is definitely ripe for review.

kohr-h commented:
I'll also look into the PEP8 issues.

kohr-h commented:
Any chance to get a review on this PR? I know it looks horribly large from the numbers, but a lot is due to movement of code. When classes go away that were contained in their own modules, it's unavoidable that their functionality move to other files. Thus you get these inflated numbers.

If it helps, I can write a small summary of how things used to be wired up and what it looks like now, so it becomes easier to digest and review.

kohr-h commented:
I'll also rebase and get rid of the PEP8 warnings.

adler-j commented:
Please try to fix the conflicts etc, then I can review this later (although perhaps not this week, it's chaos).

kohr-h commented on odl/discr/discr_utils.py:
I'll fix these weird gaps.

kohr-h commented on odl/util/utility.py:
```diff
@@ -178,18 +198,49 @@ class npy_printoptions(object):
+
+
+@contextmanager
+def npy_error_handling(**options):
```
This might actually be equivalent to `np.errstate`, can be removed in that case.

kohr-h replied:
It is, function removed.

kohr-h commented:
Would be nice to have some eyes on this PR. It mostly removes stuff, so should be a no-brainer.

**Here's a list of substantial changes:**
- `interp` is no longer a space property, goes to relevant operators instead.
  *Rationale:* Makes much more sense IMO.
- Support for non-vectorized functions in `DiscreteLp.element()` is removed.
  *Rationale:* Why did we ever support this?
- Convenience-wrapping of functions to support calling with and without `out`, as well as vector-valued functions, now happens in `make_func_for_sampling`, with the same functionality as `FunctionSpace.element()` before.
- All sampling- and interpolation-related functionality now implemented with pure functions (that's the contents of `discr_utils.py`).
- No longer use a base class for `ResizingOperator` and its adjoint, instead make an inline class for the adjoint.
- Some parts of `discretization.py` (now gone) are inlined in `lp_discr.py`, which makes it appear as if it had changed a lot. It hasn't.
- The context managers in `utility.py` are now implemented with `contextlib.contextmanager`, which makes them *much* shorter and more straightforward.
- `discr_sequence_space` is gone.
  *Rationale:* lack of usefulness.
- Instead of using a semi-broken mapping from `interp` to projector type in `astra_setup.py`, users can now explicitly specify the ASTRA projector type.

kohr-h commented on odl/discr/discr_ops.py:
```diff
@@ -73,10 +73,11 @@ def __init__(self, domain, range, interp):
+        if domain.domain != range.domain:
+            raise ValueError(
+                '`domain.domain` ({}) does not match `range.domain` ({})'
+                ''.format(domain.domain, range.domain)
```
Not sure about this error message :unamused: 
But no idea how to improve it.

kohr-h commented on odl/discr/discr_utils.py:
```diff
@@ -61,16 +72,15 @@ def point_collocation(func, points, out=None, **kwargs):
     can be switched off by passing ``bounds_check=False``:
 
-    >>> mesh = sparse_meshgrid([-1, 0, 4])
+    >>> mesh = np.meshgrid([-1, 0, 4])
```
I'll revert this

kohr-h commented on odl/discr/discr_utils.py:
```diff
@@ -817,6 +838,555 @@ def __init__(self, coord_vecs, values, input_type):
 
 
+def _check_func_out_arg(func):
+    """Check if ``func`` has an (optional) ``out`` argument.
```
Mention `func_out_type` and that this function only works for functions and `callable`s

kohr-h commented on odl/discr/discr_utils.py:
```diff
@@ -817,6 +838,555 @@ def __init__(self, coord_vecs, values, input_type):
+    else:
+        # This is for the case that an array-like of callables is provided.
+        # We need to convert this into a single function that returns an
+        # array, and maybe we need to vectorize member functions.
```
the part after the comma doesn't apply any more, remove

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -310,68 +360,38 @@ def element(self, inp=None, order=None, **kwargs):
-    # only a fraction of the outermost cells (e.g. if the outermost grid
-    # points lie at the boundary), the corresponding contribuitons to
-    # discretized integrals need to be scaled by that fraction.
-    def _inner(self, x, y):
```
Moved down, unchanged

kohr-h commented on odl/test/discr/discr_ops_test.py:
```diff
@@ -153,11 +153,15 @@ def test_resizing_op_call(odl_tspace_impl):
+        )
+        res_space = odl.uniform_discr(
+            [0, -0.6], [2, 0.2], (8, 2), dtype=dtype, impl=impl
+        )
```
Notice how we never used `dtype` before :laughing:

kohr-h commented on odl/util/utility.py:
```diff
@@ -157,11 +156,12 @@ def dedent(string, indent_str='   ', max_levels=None):
     def num_indents(line):
         max_num = int(np.ceil(len(line) / len(indent_str)))
 
+        i = 0
```
Otherwise undefined if loop does not run (`max_num == 0`)

aringh replied:
Add a comment above to clarify this?

kohr-h commented:
I've now rebased this PR on `master`. In particular the `'interp'` related stuff is already done there.

Time for review :grimacing: 

**Note:** I'm not sure how much sense it really makes to review in detail. A lot of code is just moved from `fspace.py` and `discretization.py` into `discr_utils.py` or `lp_discr.py`, respectively (the same for the corresponding tests).
I'd suggest to have a look at doctests and unit tests for the API first. Also the "unrelated" files with various small fixes.

kohr-h commented:
> Partial review. Sorry about the delay this takes a while to do. Overall it's a phenomenal change that I really like.

Thanks so far! Seems like you reviewed an outdated version of the code, thus some of the comments have already been addressed in the meanwhile. I'll have a look at the ones that are still valid.

kohr-h commented on odl/deform/linearized.py:
```diff
@@ -135,6 +149,19 @@ def __init__(self, template, domain=None):
+
+            Supported values: ``'nearest'``, ``'linear'``
+
+            .. note::
```
Do you mean `..warning::` or a `stderr` warning? I'm not a big fan of the latter since they are annoying for the "I know what I'm doing" kind of folks. And here it only affects the (admittedly rather important) case of operator derivative.
I'll go with the `..warning::` for the docs.

kohr-h commented on odl/deform/linearized.py:
```diff
@@ -226,13 +267,15 @@ def derivative(self, displacement):
     def __repr__(self):
         """Return ``repr(self)``."""
         posargs = [self.template]
+        # TODO(kohr-h): interp
```
This is done in the latest version

kohr-h commented on odl/deform/linearized.py:
```diff
@@ -285,8 +328,16 @@ def __init__(self, displacement, templ_space=None):
+            sequence gives the iterpolation scheme per axis.
+
+            Supported values: ``'nearest'``, ``'linear'``
+
```
No, this operator is linear, so there's no problem.

kohr-h commented on odl/discr/discr_ops.py:
```diff
@@ -8,88 +8,110 @@
+    normalized_scalar_param_list, resize_array, safe_int_conv, writable_array)
 from odl.util.numerics import _SUPPORTED_RESIZE_PAD_MODES
-
+from odl.util.utility import none_context
```
The whole thing is really an inferior backport of the Python 3.7+ built-in [`contextlib.nullcontext`](https://docs.python.org/3/library/contextlib.html#contextlib.nullcontext). I shied away from calling it the same because of a slightly wider API, but that should be very simple to implement.

Doing it "properly" would mean to have a `py_compat.py` module that would try to import the `contextlib` thing and fall back to a custom one if needed. We could also put the Py2 compatibility code there, e.g., `lru_cache`, `inspect.getfullargspec`, etc.

Does that sound better?

kohr-h replied:
OTOH, `lru_cache` is replaced with a dummy, and `getfullargspec` is used along with other Python version specific code, so they don't really fit into a `py_compat` module. Probably it's better to just put the code in `utility.py` for now.

kohr-h commented on odl/discr/discr_utils.py:
```diff
@@ -0,0 +1,1432 @@
+
+
+def _all_interp_equal(interp_byaxis):
+    """Whether all entries are equal, with ``False`` for length 0."""
```
It's not super beautiful, but its only purpose is for the `interp` property to decide whether it should print a single string or a tuple of strings. For length 0, it has to print a tuple, so we need a `False` in that case.
Maybe the name is bad. How about `_is_single_interp(interp_byaxis)`?

kohr-h replied:
I'll just remove the confusing helper and inline the code.

kohr-h commented on odl/discr/discr_utils.py:
```diff
@@ -0,0 +1,1432 @@
+    """
+    f = np.asarray(f)
+
+    # TODO(kohr-h): pass reasonable options on to the interpolator
```
No, this is an old comment. Not sure if there's really anything to do here.

kohr-h commented on odl/discr/discr_utils.py:
```diff
@@ -0,0 +1,1432 @@
+
+
+def linear_interpolator(f, coord_vecs):
+    """Return the linear interpolator for discrete values.
```
"discrete function values" I guess

kohr-h commented on odl/discr/discr_utils.py:
```diff
@@ -0,0 +1,1432 @@
+                raise ValueError('`nn_variants` {!r} not understood'
+                                 ''.format(nn_variants_in))
+        else:
+            nn_variants = [str(var).lower() if var is not None else None
```
Removed in latest version

kohr-h commented on odl/discr/discr_utils.py:
```diff
@@ -0,0 +1,1432 @@
+        # sequences of inputs, single-input case has been checked above
+        if schemes[i] not in _SUPPORTED_INTERP_SCHEMES:
+            raise ValueError('`interp[{}]={!r}` not understood'
+                             ''.format(schemes_in[i], i))
```
Code removed

kohr-h commented on odl/discr/discr_utils.py:
```diff
@@ -0,0 +1,1432 @@
+            raise ValueError('`input_type` ({}) not understood'
+                             ''.format(input_type))
+
+        if len(coord_vecs) > values.ndim:
```
I don't think so. Possibly this code was copied from the `scipy` template class. I'll change it.

kohr-h commented on odl/discr/discr_utils.py:
```diff
@@ -0,0 +1,1432 @@
+    return w_lo, w_hi, edge
+
+
+def _compute_linear_weights_edge(idcs, ndist):
```
Yeah, no code change here

kohr-h commented on odl/discr/discr_utils.py:
```diff
@@ -0,0 +1,1432 @@
+        self.nn_variants = nn_variants
+
+    def _evaluate(self, indices, norm_distances, out=None):
+        """Evaluate linear interpolation.
```
Of course per-axis

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -73,55 +62,23 @@ def __init__(self, fspace, partition, tspace, interp='nearest', **kwargs):
-                fspace, self.partition, tspace, self.interp_byaxis)
-
-        super(DiscreteLp, self).__init__(fspace, tspace, sampling, interpol)
+        # TODO: will change
```
Done?!

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -493,33 +446,112 @@ def __getitem__(self, indices):
+
+    def __hash__(self):
+        """Return ``hash(self)``."""
+        return hash((super(DiscreteLp, self).__hash__(), self.tspace))
```
Oh yes, good catch

kohr-h commented on odl/discr/discr_ops.py:
```diff
@@ -36,10 +35,11 @@ def __init__(self, domain, range, interp):
         Parameters
         ----------
-        domain : `DiscretizedSpace`
+        domain : `DiscreteLp`
```
Hm, I was skeptical on first thought, but on second thought, since this space is so ubiquitous we can as well simplify the name. I'll go for `DiscretizedSpace`. :+1:

kohr-h commented on odl/discr/discr_utils.py:
```diff
@@ -6,23 +6,34 @@
     'nearest_interpolator',
     'linear_interpolator',
     'per_axis_interpolator',
+    'make_func_for_sampling',
```
Okay, I'll go for `sampling_function`

kohr-h commented on odl/discr/discr_utils.py:
```diff
@@ -106,14 +115,14 @@ def point_collocation(func, points, out=None, **kwargs):
     >>> def vec_valued(x):
     ...     return (x[0] - 1, 0, x[0] + x[1])  # broadcasting
-    >>> func1 = fspace.element(vec_valued)
+    >>> # We must tell the wrapper that we want a 3-component function
```
This is no regression. In the old version it just happened earlier in the doctest.

kohr-h commented on odl/discr/discr_utils.py:
```diff
@@ -123,12 +132,13 @@ def point_collocation(func, points, out=None, **kwargs):
     ... ]
-    >>> func2 = fspace.element(list_of_funcs)
+    >>> func2 = make_func_for_sampling(
+    ...     array_of_funcs, domain, out_dtype=(float, (3,)))
```
In principle yes, I'll look into it.

kohr-h replied:
I've made a small change so that for an array of functions, the shape is inferred.

kohr-h commented on odl/discr/discr_utils.py:
```diff
@@ -560,8 +569,15 @@ def __call__(self, x, out=None):
         """
+        x = np.asarray(x)
         ndim = len(self.coord_vecs)
+        scalar_out = False
```
Not sure what you mean here

kohr-h commented on odl/discr/discr_utils.py:
```diff
@@ -817,6 +838,559 @@ def __init__(self, coord_vecs, values, input_type):
+    Parameters
+    ----------
+    func_or_arr : callable or array-like
+        Either a callable object or an array or callables and constants.
```
Added a bit of text

kohr-h commented on odl/discr/discr_utils.py:
```diff
@@ -123,12 +132,13 @@ def point_collocation(func, points, out=None, **kwargs):
            [[ 4.,  5.],
             [ 5.,  6.]]])
-    >>> list_of_funcs = [
+    >>> array_of_funcs = [  # equivalent to `vec_valued`
```
Done

kohr-h commented on odl/discr/discr_utils.py:
```diff
@@ -0,0 +1,1432 @@
+    - The distinction between left and right variants is currently
+      made by changing ``<=`` to ``<`` at one place. This difference
+      may not be noticable in some situations due to rounding errors.
+
```
Text is gone

kohr-h commented on README.md:
```diff
@@ -12,7 +12,6 @@ For more details and an introduction into the inner workings of ODL, please refe
 
 Highlights
 ==========
-- Well-tested data containers based on NumPy or CUDA allow high performance computing with minimal effort.
```
Weird that we still had this on the front page README

adler-j replied:
Probably should have been in another PR but sure :)

kohr-h commented on doc/source/getting_started/about_odl.rst:
```diff
@@ -16,182 +16,6 @@ The main advantages of this approach is that
 4. Adding new applications or solution methods becomes a much easier task.
 
-ODL implements many abstract mathematical notions such as sets, vector spaces and operators.
-In the following, a few are shown by example.
```
All the stuff below seemed utterly pointless to me. What normal user cares about this stuff?

adler-j replied:
I agree, go ahead.

kohr-h commented:
Done with the first rounds of review comments. I changed the class names as suggested, and I think it makes a lot of sense like this.
How about the rest @adler-j ? :wave:

kohr-h commented:
Hey, ping again. I feel we're almost there with this one. Reviewers, please try to give it a last push (or click "approve" :-) ).

aringh commented on odl/test/trafos/fourier_test.py:
```diff
@@ -46,6 +46,20 @@ def _params_from_dtype(dtype):
     return halfcomplex, complex_dtype(dtype)
 
 
+def _dft_space(shape, dtype='float64'):
```
This seems to replaces ``odl.discr_sequence_space``. Was the other one used only here? Is there no reason to keep such a space/helper somewhere?

Edit: in ``odl/trafos/fourier``, the sequence space seems to have been replaced by uniform discretization. Can this be done here too?

kohr-h replied:
I've basically degraded `discr_sequence_space` to a test helper and inlined the code in `fourier.py` where it is used twice. The main reason for doing that is that for its name "sequence space" it was just a hack using a discrete function space which happened to yield the same results.
If we ever feel the need to make such a helper available we could revive it somewhere in `odl.util`, but for an "official" discrete space API, I think it was too hacky and not useful enough.

aringh commented on odl/trafos/fourier.py:
```diff
@@ -385,15 +387,15 @@ def __init__(self, domain, range=None, axes=None, sign='-',
             is determined from ``domain`` and the other parameters as
-            a `discr_sequence_space` with exponent ``p / (p - 1)``
-            (read as 'inf' for p=1 and 1 for p='inf').
+            a `uniform_discr` with exponent unit cell size and exponent
```
Remove first ``exponent``.

kohr-h replied:
```suggestion
            a `uniform_discr` with unit cell size and exponent
```

aringh commented on odl/trafos/fourier.py:
```diff
@@ -536,15 +540,15 @@ def __init__(self, range, domain=None, axes=None, sign='+',
-            as a `discr_sequence_space` with exponent ``p / (p - 1)``
-            (read as 'inf' for p=1 and 1 for p='inf').
+            domain is determined from ``range`` and the other parameters as
+            a `uniform_discr` with exponent unit cell size and exponent
```
Same as above.

kohr-h replied:
```suggestion
            a `uniform_discr` with unit cell size and exponent
```

aringh commented on odl/discr/lp_discr.py:
```diff
@@ -104,19 +94,61 @@ def __init__(self, fspace, partition, tspace, **kwargs):
     @property
     def exponent(self):
         """Exponent of this space, the ``p`` in ``L^p``."""
+        # TODO(kohr-h): `exponent` is optional in `tspace`, how should we
```
Isn't this important? What happens if the underlying tensor space has no exponent, but I want my discretized space to have one? Or is the philosophy (at least at this moment) to simply leave it to the user to make sure that things are consistent?

kohr-h replied:
Yes, it is important. But the paradigm here is that the exponent is simply whatever the exponent of the wrapped `tspace` is. If it doesn't exist, then the `AttributeError` will just bubble up. I'm not sure how we can do better than that.

kohr-h commented:
> Final note: `odl/discr/discr_utils` was long and somewhat difficult to understand; I will be the first to admit that I did not follow all the details. But I guess most of it is comming directly from the `odl/space/fspace` file, that was deleted, right? Moreover, I have actually not checked `odl/test/discr/discr_utils_test`, but I guess most of it is moved from `odl/test/space/fspace_test`.

That's exactly right. The tests are really just copy-pasted to the new file, and also the `discr_utils.py` module largely consists of the old contents of `fspace.py` and `discr_mappings.py` contents, maybe slightly streamlined in some parts. All the old tests are still performed, so what worked before still works.

kohr-h commented on doc/source/guide/glossary.rst:
```diff
@@ -21,93 +18,74 @@ Glossary
-        points ``(1, -1), (1, 0), (1, 1), (2, -1), (2, 0), (2, 1)``.
+        Tuple of arrays defining a tensor grid by all possible combinations of entries, one from each array.
+        In 2 dimensions, for example, the arrays ``[[1], [2]]`` and ``[[-1, 0, 1]]`` define the grid points ``(1, -1), (1, 0), (1, 1), (2, -1), (2, 0), (2, 1)``.
+        Note that the resulting grid has the broadcast shape, here ``(2, 3)``, broadcast from ``(2, 1)`` and ``(1, 3``.)
```
In code it would be `result_shape = np.broadcast(shape1, shape2).shape`. (And I see a typo with the last closing paren.)

kohr-h replied:
```suggestion
        Note that the resulting grid has the broadcast shape, here ``(2, 3)``, broadcast from ``(2, 1)`` and ``(1, 3)``
        (expressed in code: ``result_shape = np.broadcast(shape1, shape2).shape``).
```

kohr-h commented on odl/deform/linearized.py:
```diff
@@ -231,7 +230,9 @@ def interp_byaxis(self):
     def interp(self):
         """Interpolation scheme or tuple of per-axis interpolation schemes."""
-        if _all_interp_equal(self.interp_byaxis):
+        if len(self.interp_byaxis) == 0:
```
If we didn't have this special case, then it would fall under the next one (because `all(empty collection)` evaluates to `True`), and then fail in the indexing operation.
But maybe it's simpler to just add a `len(self.interp_byaxis) != 0 and` to that condition and handle the empty case in the `else`.

kohr-h commented on odl/discr/discr_utils.py:
```diff
@@ -817,6 +834,573 @@ def __init__(self, coord_vecs, values, input_type):
+        return out
+
+    # Now prepare the in-place and out-of-place functions for the final
+    # wrapping. If needed, vectorize.
```
Last sentence obsolete

kohr-h replied:
```suggestion
    # wrapping.
```

kohr-h commented on odl/discr/discr_utils.py:
```diff
@@ -817,6 +834,573 @@ def __init__(self, coord_vecs, values, input_type):
+            elif is_valid_input_array(x, domain.ndim):
+                scalar_out_shape = out_shape_from_array(x)
+            else:
+                raise RuntimeError('bad input')
```
Hm, yeah, you're right. I think there's an error message somewhere for the same situation. I'll try to dig it out.

kohr-h replied:
I think I could generalize [this function](https://github.com/odlgroup/odl/blob/b017445809567588f69828064ef816e5562c0229/odl/discr/discr_utils.py#L209-L249)

kohr-h replied:
Actually that won't be necessary since the check is always done [here](https://github.com/odlgroup/odl/blob/b017445809567588f69828064ef816e5562c0229/odl/discr/discr_utils.py#L1226-L1253). So at this point we should not get any error anymore, and that scenario would indeed be a `RuntimeError`.

kohr-h commented on odl/util/utility.py:
```diff
@@ -157,11 +150,12 @@ def dedent(string, indent_str='   ', max_levels=None):
     def num_indents(line):
         max_num = int(np.ceil(len(line) / len(indent_str)))
 
+        i = 0
```
```suggestion
        i = 0  # set for the case the loop is not run (`max_num == 0`)
```

kohr-h commented on odl/deform/linearized.py:
```diff
@@ -231,7 +230,9 @@ def interp_byaxis(self):
-        if _all_interp_equal(self.interp_byaxis):
+        if len(self.interp_byaxis) == 0:
+            return ()
+        elif all(s == self.interp_byaxis[0] for s in self.interp_byaxis[1:]):
```
```suggestion
        if (
            len(self.interp_byaxis) != 0
            and all(s == self.interp_byaxis[0] for s in self.interp_byaxis[1:])
        ):
```

kohr-h commented on odl/deform/linearized.py:
```diff
@@ -409,7 +410,9 @@ def interp_byaxis(self):
-        if _all_interp_equal(self.interp_byaxis):
+        if len(self.interp_byaxis) == 0:
+            return ()
+        elif all(s == self.interp_byaxis[0] for s in self.interp_byaxis[1:]):
```
```suggestion
        if (
            len(self.interp_byaxis) != 0
            and all(s == self.interp_byaxis[0] for s in self.interp_byaxis[1:])
        ):
```

kohr-h commented on odl/discr/discr_ops.py:
```diff
@@ -91,7 +92,9 @@ def interp_byaxis(self):
-        if _all_interp_equal(self.interp_byaxis):
+        if len(self.interp_byaxis) == 0:
+            return ()
+        elif all(s == self.interp_byaxis[0] for s in self.interp_byaxis[1:]):
```
```suggestion
        if (
            len(self.interp_byaxis) != 0
            and all(s == self.interp_byaxis[0] for s in self.interp_byaxis[1:])
        ):
```

kohr-h commented on odl/discr/discr_utils.py:
```diff
@@ -817,6 +834,573 @@ def __init__(self, coord_vecs, values, input_type):
+    ----------
+    func_or_arr : callable or array-like
+        Either a single callable object (possibly with multiple output
+        components), or an array or callables and constants.
```
```suggestion
        components), or an array or callables and constants.
        A callable (or each callable) must take a single input and may
        accept one output parameter called ``out``, and should return
        its result.
```
To address [this comment](https://github.com/odlgroup/odl/pull/1459/files/eaebdc6bf721345bc1614a756ed3e12264b3fbe0#r400704096)

kohr-h commented on odl/discr/discr_utils.py:
```diff
@@ -817,6 +834,573 @@ def __init__(self, coord_vecs, values, input_type):
+                'ufunc {} returns {} outputs, expected 0 or 1'
+                ''.format(func.__name__, func.nout)
+            )
+        has_out = out_optional = (func.nout == 1)
```
See [below](https://github.com/odlgroup/odl/pull/1459/files/eaebdc6bf721345bc1614a756ed3e12264b3fbe0#r400786698)

kohr-h commented on odl/contrib/fom/supervised.py:
```diff
@@ -828,9 +828,9 @@ def noise_power_spectrum(data, ground_truth, radial=False,
+    data : `DiscretizedSpaceElement` or `array-like`
         Input data to compare to the ground truth. If not a
-        `DiscreteLpElement`, a default space with cell size 1 will be assumed.
+        `DiscretizedSpaceElement`, a default space with cell size 1 will be assumed.
```
```suggestion
        `DiscretizedSpaceElement`, a default space with cell size 1 will be
        assumed.
```

kohr-h commented on odl/deform/linearized.py:
```diff
@@ -375,18 +377,18 @@ def __init__(self, displacement, templ_space=None, interp='linear'):
             raise ValueError(
-                '`displacement.space[0]` must be a `DiscreteLp`, got {!r}'
+                '`displacement.space[0]` must be a `DiscretizedSpace`, got {!r}'
                 ''.format(displacement.space[0]))
```
```suggestion
                '`displacement.space[0]` must be a `DiscretizedSpace`, '
                'got {!r}'.format(displacement.space[0]))
```

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -1118,9 +1303,7 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
-                    res_space = DiscreteLp(
-                        fspace, part, tspace, axis_labels=labels
-                    )
+                    res_space = DiscretizedSpace(part, tspace, axis_labels=labels)
```
```suggestion
                    res_space = DiscretizedSpace(
                        part, tspace, axis_labels=labels
                    )
```

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -1076,27 +1269,19 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
 
             if out is None:
-                # Wrap in appropriate DiscreteLp space depending on `method`
+                # Wrap in appropriate DiscretizedSpace space depending on `method`
```
```suggestion
                # Wrap in appropriate DiscretizedSpace space depending
                # on `method`
```

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -943,7 +1153,7 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
         # --- Process `inputs` --- #
 
-        # Pull out the `tensor` attributes from DiscreteLpElement instances
+        # Pull out the `tensor` attributes from DiscretizedSpaceElement instances
```
```suggestion
        # Pull out the `tensor` attributes from `DiscretizedSpaceElement`
        # instances
```

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -445,29 +467,115 @@ def __getitem__(self, indices):
+    def __hash__(self):
+        """Return ``hash(self)``."""
+        return hash(
+            (super(DiscretizedSpace, self).__hash__(), self.tspace, self.partition)
```
```suggestion
            (super(DiscretizedSpace, self).__hash__(),
            self.tspace,
            self.partition)
```

kohr-h commented on odl/test/discr/lp_discr_test.py:
```diff
@@ -284,16 +274,12 @@ def f(x, **kwargs):
 
 def test_element_from_function_2d():
-    """Test creation of DiscreteLp elements from functions in 2 dimensions."""
+    """Test creation of DiscretizedSpace elements from functions in 2 dimensions."""
```
```suggestion
    """Test creation of DiscretizedSpace elements from functions in 2D."""
```

kohr-h commented on odl/test/discr/lp_discr_test.py:
```diff
@@ -244,7 +234,7 @@ def test_element_from_array_2d(odl_elem_order):
 
 def test_element_from_function_1d():
-    """Test creation of DiscreteLp elements from functions in 1 dimension."""
+    """Test creation of DiscretizedSpace elements from functions in 1 dimension."""
```
```suggestion
    """Test creation of DiscretizedSpace elements from functions in 1D."""
```

kohr-h commented on odl/trafos/fourier.py:
```diff
@@ -79,11 +79,11 @@ def __init__(self, inverse, domain, range=None, axes=None, sign='-',
-            raise TypeError('`domain` {!r} is not a `DiscreteLp` instance'
+        if not isinstance(domain, DiscretizedSpace):
+            raise TypeError('`domain` {!r} is not a `DiscretizedSpace` instance'
                             ''.format(domain))
```
```suggestion
            raise TypeError(
                '`domain` {!r} is not a `DiscretizedSpace` instance'
                ''.format(domain)
            )
```

kohr-h commented on odl/trafos/wavelet.py:
```diff
@@ -139,8 +139,8 @@ def __init__(self, space, wavelet, nlevels, variant, pad_mode='constant',
-            raise TypeError('`space` {!r} is not a `DiscreteLp` instance.'
+        if not isinstance(space, DiscretizedSpace):
+            raise TypeError('`space` {!r} is not a `DiscretizedSpace` instance.'
                             ''.format(space))
```
```suggestion
            raise TypeError(
                '`space` {!r} is not a `DiscretizedSpace` instance.'
                ''.format(space)
            )
```

kohr-h commented:
All comments addressed and rebased on master. Unless there are complaints (by a reviewer or by CI) I'll merge later today.

kohr-h commented:
One last thing I did was to rename the `lp_discr` module to `discr_space`, in line with the rename of `DiscreteLp`. I'll wait for CI and then merge.

adler-j commented:
Very well done on getting this in Holger!
